### PR TITLE
feat(core): implement audit module

### PR DESCRIPTION
**Pull request by eirikurn**
## What

Nest module to audit user actions in our backends.

Includes:

* New `@island.is/nest/audit` module.
* Refactored auth-nest-utils, to better handle different kinds of auth guards, and forwarding of credentials.
* Change from IdsAuthGuard to IdsUserGuard in most places.
* Audit integration in Application System

## Why

So we can audit who did what. This is especially important when we launch the delegation system which allows users to sign in on behalf of other users and companies.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: apps/application-system/api/src/environments/environment.interface.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import { TemplateAPIConfig } from '@island.is/application/template-api-modules'
 import { ApplicationConfig } from '../app/modules/application/application.configuration'
 import { FileStorageConfig } from '@island.is/file-storage'
 import { SigningServiceOptions } from '@island.is/dokobit-signing'
-import { SyslumennClientConfig } from '@island.is/api/domains/syslumenn'
+import { AuditOptions } from '@island.is/nest/audit'
```
Comment (jeremybarbet) [inline apps/application-system/api/src/environments/environment.interface.ts:6]: This config is not necesary anymore? `import { SyslumennClientConfig } from '@island.is/api/domains/syslumenn'`

**Review by jeremybarbet — COMMENTED** [review]
Only a minor comment, apart from that it looks good to me

**Review by valurefugl — APPROVED** [review]

File: apps/application-system/api/src/environments/environment.interface.ts (hunk 1 above)
Comment (eirikurn) [reply apps/application-system/api/src/environments/environment.interface.ts:6]: It’s not used. It’s part of the template api options instead.

File: libs/auth-nest-tools/src/lib/current-auth.decorator.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import { getRequest } from './getRequest'
+
+export const getCurrentAuth = (context: ExecutionContext): Auth => {
+  const request = getRequest(context)
+
+  const auth = request.auth
+  if (!auth) {
+    console.warn(
+      'No authentication found. Did you forget to add IdsAuthGuard or IdsUserGuard?',
+    )
```
Comment (stefanhar) [inline libs/auth-nest-tools/src/lib/current-auth.decorator.ts:16]: Is logging to console intentional here or should this maybe use the logger?

File: libs/auth-nest-tools/src/lib/current-user.decorator.ts [hunk 3]
```diff
@@ -3,37 +3,25 @@ import {
...
 
+  const user = request.user
   if (!user) {
-    throw new UnauthorizedException('You are not authenticated')
-  }
-
-  return {
-    ...user,
-    authorization: request.headers.authorization,
+    console.warn(
```
Comment (kksteini) [inline libs/auth-nest-tools/src/lib/current-user.decorator.ts:15]: Is logging to console done on purpose?

File: libs/nest/README.md [hunk 4]
```diff
@@ -0,0 +1 @@
+# Nest Modules
```
Comment (kksteini) [inline libs/nest/README.md:1]: Is this README supposed to be just a heading?

File: libs/nest/audit/src/lib/audit.interceptor.spec.ts [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+  it('forwards user', async () => {
+    // Arrange
+    class MyClass {
+      @Audit()
+      handler() {}
+    }
+    context.getClass.mockReturnValue(MyClass)
+    context.getHandler.mockReturnValue(MyClass.prototype.handler)
+    const user = 'Test user'
+    ;((getCurrentUser as unknown) as MockInstance<
```
Comment (kksteini) [inline libs/nest/audit/src/lib/audit.interceptor.spec.ts:127]: Is this: https://prettier.io/docs/en/rationale.html#semicolons what's happening here?

**Review by kksteini — APPROVED** [review]

File: libs/nest/audit/src/lib/audit.interceptor.spec.ts (hunk 5 above)
Comment (eirikurn) [reply libs/nest/audit/src/lib/audit.interceptor.spec.ts:127]: Yeah... :( Will rewrite as variable.

File: libs/nest/README.md (hunk 4 above)
Comment (eirikurn) [reply libs/nest/README.md:1]: Yeah, that's the (lazy) current pattern for these middle folders, which are required for our gitbook scripts.

File: libs/auth-nest-tools/src/lib/current-user.decorator.ts (hunk 3 above)
Comment (eirikurn) [reply libs/auth-nest-tools/src/lib/current-user.decorator.ts:15]: Nope! Should be using the common logger. Good catch!

File: libs/auth-nest-tools/src/lib/current-auth.decorator.ts (hunk 2 above)
Comment (eirikurn) [reply libs/auth-nest-tools/src/lib/current-auth.decorator.ts:16]: Good catch!

**Review by fjandakornid — APPROVED** [review]
LGTM

**Review by dd-sdang — APPROVED** [review]
LGMT

**Review by jeremybarbet — APPROVED** [review]
LGTM 👍

File: apps/application-system/api/src/app/modules/application/application.controller.ts [hunk 6]
```diff
@@ -157,19 +164,27 @@ export class ApplicationController {
...
   async findAll(
-    @CurrentRestUser() user: User,
+    @Param('nationalId') nationalId: string,
+    @CurrentUser() user: User,
     @Query('typeId') typeId?: string,
     @Query('status') status?: string,
   ): Promise<ApplicationResponseDto[]> {
+    if (nationalId !== user.nationalId) {
+      throw new UnauthorizedException()
+    }
```
Comment (pshomov) [inline apps/application-system/api/src/app/modules/application/application.controller.ts:178]: Suggestion for improvement: this to be extracted as part of a guard ... somehow ;)

File: apps/application-system/api/src/app/modules/application/e2e/application.spec.ts [hunk 7]
```diff
@@ -60,20 +61,16 @@ beforeAll(async () => {
...
+        .overrideGuard(IdsUserGuard)
+        .useValue(
+          new MockAuthGuard({
+            nationalId,
+            scope: [
+              ApplicationIdentityServerScope.read,
+              ApplicationIdentityServerScope.write,
+            ],
+          }),
+        )
```
Comment (pshomov) [inline apps/application-system/api/src/app/modules/application/e2e/application.spec.ts:73]: Very nice 👍🏻

**Review by pshomov — APPROVED** [review]
LGTM,

Since I am only looking at the diff and it is quite late: I am not seeing tests about the auth decorators/guard things and did not see them refactored in unit tests. Maybe I am missing something but if I am not - we should add a few test cases to see that we handle strange/malformed tokens if possible.

File: apps/application-system/api/src/app/modules/application/application.controller.ts (hunk 6 above)
Comment (pshomov) [reply apps/application-system/api/src/app/modules/application/application.controller.ts:178]: Future improvement I meant ;)

**Review by stefanhar — APPROVED** [review]

**Pull request merged**